raiden/utils/gas_escrow.py

```diff
@@ -0,0 +1,93 @@
+from typing import Tuple
+
+from raiden.constants import (
+    GAS_USED_OPEN_CHANNEL,
+    GAS_USED_SET_TOTAL_DEPOSIT,
+    GAS_USED_CLOSE_CHANNEL,
+    GAS_USED_SETTLE_CHANNEL,
+    GAS_USED_UNLOCK_1_LOCKS,
+)
+from raiden.transfer import views
+
+CHANNEL_LIFECYCLE_SETTLE_GAS_USED = GAS_USED_SETTLE_CHANNEL + GAS_USED_UNLOCK_1_LOCKS
```

Comment:
Good point. Question is how conservative we want to be. `MAXIMUM_PENDING_TRANSFERS` seems like overkill in most cases but is a proper upper bound.

Reply:
To me conservative is worst case.

I understand the point of this is to show the user a warning when exiting will not be possible because of gas. If these two constants are *not coherent with each* then the warning would not be shown even if the *channel* cannot be properly settled.

Also `MAXIMUM_PENDING_TRANSFERS` is computed out of an arbitrary gas limit, which is [40% of the block gas](https://github.com/raiden-network/raiden/pull/1990/files#diff-8d42ec50aee7c21051d55b0748bc45f2R219), using a lower number to reduce the required gas here should not be a problem.

Reply:
The maximum number of pending locks is not being take care off. @LefterisJP you don't want to have the worst case scenario covered?

Reply:
@hackaugusto @palango  I do want to have the worst case scenario covered. Please "Request changes" to signify to other reviewers (in this case myself) that you have a blocking request or else since github marks your comment as "Outdated" I assume it's fixed.

But then I got a question:

> Note that the gas for the maximum number of unlocks is not GAS_USED_UNLOCK_1_LOCKS * MAXIMUM_PENDING_TRANSFERS

So what is it if not that? I guess ofcourse that's multiplied by the number of open channels but I guess this is not what you meant.

Reply:
> So what is it if not that?

It's the arbitrary value chosen on issue https://github.com/raiden-network/raiden/issues/438 , which should be exposed by the PR https://github.com/raiden-network/raiden/pull/1990/files#r206983562

Reply:
I'll create an issue for follow up.

Reply:
Aha, so the actual value may be incorrect due to the arbitrary nature of the 40%. But assuming we find a good value for `MAXIMUM_PENDING_TRANSFERS` then the calculation would be correct.

Reply:
> I'll create an issue for follow up.

No need, I will push an extra commit to jochen's PR when he rebases it on master. It's too small for an issue.

Reply:
done: https://github.com/raiden-network/raiden/pull/1990/commits/814463fd5f66df69854ad21d325f3664d8fbefd6

Let's take the conversation of whether 40% of the block gas is a lot in that PR.

Reply:
> But assuming we find a good value for MAXIMUM_PENDING_TRANSFERS then the calculation would be correct.

No, that is not how it works, `GAS_USED_UNLOCK_1_LOCKS * MAXIMUM_PENDING_TRANSFERS` is always wrong.

`GAS_USED_UNLOCK_1_LOCKS` is the gas cost for **one** transaction that unlocks one lock, if you multiply it by `MAXIMUM_PENDING_TRANSFERS` you will have the gas cost of `n` **transactions**, which is completely different from *one* transaction which unlocks `n` locks.

The most obvious difference is that sending `n` transaction will have to pay the minimum `21000` gas cost `n` times.

Reply:
Hmm @hackaugusto but I am assuming a batch unlock of all the locks together not a different transaction for each.

Reply:
> Hmm @hackaugusto but I am assuming a batch unlock of all the locks together not a different transaction for each.

Why are you using `GAS_USED_UNLOCK_1_LOCKS * MAXIMUM_PENDING_TRANSFERS` then? That expression computes `n` transactions, and not 1 transaction for "batch unlock of all the locks".

Reply:
Well technically you can still provide them one by one but since we are the only ones who call it, then it's safe to say we should only consider batch unlock. Then computing the proper amount of gas may be a bit trickier. 

But the end result should be to compute how much it costs in gas to unlock `MAXIMUM_PENDING_TRANSFERS` locks in one transaction. @hackaugusto are we in agreement for this?

Reply:
> Well technically you can still provide them one by one but since we are the only ones who call it

No, we cannot, there is not such function anymore. [The whole merkle tree must be provided](https://github.com/raiden-network/raiden-contracts/blob/fff8646ebcf2c812f40891c2825e12ed03cc7628/raiden_contracts/contracts/TokenNetwork.sol#L705-L710)

> But the end result should be to compute how much it costs in gas to unlock MAXIMUM_PENDING_TRANSFERS locks in one transaction

Yes, this is the constant that I mentioned [here](https://github.com/raiden-network/raiden/pull/2031#discussion_r207895366) and [which is defined here](https://github.com/raiden-network/raiden/pull/1990/files#diff-fc15643f6fbeaa49665008430fe239bbR36)

Reply:
>  The whole merkle tree must be provided

Argh ... ofcourse. Brainfart .. 

> Yes, this is the constant that I mentioned here and which is defined here

Cool, we are in sync then. The commit will be amended.